Source/Core/Core/HW/WiimoteEmu/WiimoteEmu.h

```diff
@@ -205,6 +210,127 @@ enum
   ACCEL_RANGE = (ACCEL_ONE_G - ACCEL_ZERO_G),
 };
 
+class I2CSlave
+{
+public:
+  // Kill MSVC warning:
+  virtual ~I2CSlave() = default;
+
+  virtual int BusRead(u8 slave_addr, u8 addr, int count, u8* data_out) = 0;
+  virtual int BusWrite(u8 slave_addr, u8 addr, int count, const u8* data_in) = 0;
+
+  template <typename T>
+  static int RawRead(T* reg_data, u8 addr, int count, u8* data_out)
+  {
+    static_assert(std::is_pod<T>::value);
+    static_assert(0x100 == sizeof(T));
+
+    // TODO: addr wraps around after 0xff
+
+    u8* src = reinterpret_cast<u8*>(reg_data) + addr;
+    count = std::min(count, int(reinterpret_cast<u8*>(reg_data + 1) - src));
+
+    std::copy_n(src, count, data_out);
+
+    return count;
+  }
+
+  template <typename T>
+  static int RawWrite(T* reg_data, u8 addr, int count, const u8* data_in)
+  {
+    static_assert(std::is_pod<T>::value);
+    static_assert(0x100 == sizeof(T));
+
+    // TODO: addr wraps around after 0xff
+
+    u8* dst = reinterpret_cast<u8*>(reg_data) + addr;
+    count = std::min(count, int(reinterpret_cast<u8*>(reg_data + 1) - dst));
+
+    std::copy_n(data_in, count, dst);
+
+    return count;
+  }
+};
+
+class I2CBus
+{
+public:
+  void AddSlave(I2CSlave* slave) { m_slaves.emplace_back(slave); }
+
+  void RemoveSlave(I2CSlave* slave)
+  {
+    m_slaves.erase(std::remove(m_slaves.begin(), m_slaves.end(), slave), m_slaves.end());
+  }
+
+  void Reset() { m_slaves.clear(); }
+
+  int BusRead(u8 slave_addr, u8 addr, int count, u8* data_out)
+  {
+    INFO_LOG(WIIMOTE, "i2c bus read: 0x%02x @ 0x%02x (%d)", slave_addr, addr, count);
+    for (auto& slave : m_slaves)
+    {
+      auto const bytes_read = slave->BusRead(slave_addr, addr, count, data_out);
+
+      // A slave responded, we are done.
+      if (bytes_read)
+        return bytes_read;
+    }
+
+    return 0;
+  }
+
+  int BusWrite(u8 slave_addr, u8 addr, int count, const u8* data_in)
+  {
+    INFO_LOG(WIIMOTE, "i2c bus write: 0x%02x @ 0x%02x (%d)", slave_addr, addr, count);
+    for (auto& slave : m_slaves)
+    {
+      auto const bytes_written = slave->BusWrite(slave_addr, addr, count, data_in);
+
+      // A slave responded, we are done.
+      if (bytes_written)
+        return bytes_written;
+    }
+
+    return 0;
+  }
+
+private:
+  std::vector<I2CSlave*> m_slaves;
```

Comment:
I wonder if this should be a `unique_ptr` (or `shared_ptr`) instead?

Reply:
I don't use any dynamic memory allocation. It's just pointers to normally defined members. No deleting needed.